packages/flutter/lib/src/widgets/automatic_keep_alive.dart

```diff
@@ -137,16 +150,33 @@ class _AutomaticKeepAliveState extends State<AutomaticKeepAlive> {
           //    information, but it may or may not get applied this frame,
           //    depending on whether said child is in the same layout scope.
           //
+          //  * A descendant is being moved from one position under us to
+          //    another position under us. They just notified us of the removal,
+          //    at some point in the future they will notify us of the addition.
+          //    We don't want to do anything. (This is why we check that
+          //    _handles is still empty below.)
+          //
           //  * We're being notified in the paint phase, or even in a post-frame
           //    callback. Either way it is far too late for us to make our
           //    parent lay out again this frame, so the garbage won't get
           //    collected this frame.
           //
+          //  * We are being torn out of the tree ourselves, as is our
+          //    descendant, and it notified us while it was being deactivated.
+          //    We don't need to do anything, but we don't know yet because we
+          //    haven't been deactivated yet. (This is why we check mounted
+          //    below before calling setState.)
+          //
           // Long story short, we have to schedule a new frame and request a
           // frame there, but this is generally a bad practice, and you should
           // avoid it if possible.
+          _keepingAlive = false;
           scheduleMicrotask(() {
-            setState(() { _keepingAlive = false; });
+            if (mounted && _handles.isEmpty) {
+              setState(() {
+                assert(!_keepingAlive);
+              });
+            }
```

Comment:
That looks better.  It's too bad micotasks aren't cancellable otherwise you could clean this up a bit by cancelling the deferred work in cases where it becomes unnecessary.  This is fine too.